src/ast.rs

```diff
@@ -337,14 +337,17 @@ fn typ(db: &dyn Ast, decl: Arc<Decl>) -> Option<Arc<Decl>> {
                 1 => db.resolve_type((**id).clone(), loc),
                 _ => None,
             },
-            Type::Set(xs) => {
-                // TODO(bbannier): Right now we only resolve indices over sets of one key.
-                if xs.len() == 1 {
-                    xs.first().and_then(|x| db.resolve_type(x.clone(), loc))
-                } else {
-                    None
+            Type::Set(xs) => match xs.len() {
+                0 => None,
+                1 => xs.first().and_then(|x| db.resolve_type(x.clone(), loc)),
+                _ => {
+                    if *i < xs.len() {
+                        db.resolve_type(xs[*i].clone(), loc)
+                    } else {
+                        None
+                    }
                 }
-            }
+            },
```

Comment:
You need to do a lot of index checking here since indexing via `[..]` would panic on invalid indices (looks like you handle this well). A simpler way would be to [`get`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get) which would return `None` for absent elements; you can use the normal monadic functions of `Option<T>` like [`map`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map) or [`and_then`](https://doc.rust-lang.org/std/option/enum.Option.html#method.and_then), e.g., for this whole block

```suggestion
            Type::Set(xs) => xs.get(*i).and_then(|x| db.resolve_type(x.clone(), loc)),
```